staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go

```diff
@@ -0,0 +1,210 @@
+		return Validations{Comments: []string{"optional non-pointer structs are purely documentation"}}, nil
+	}
+	return Validations{Functions: []FunctionGen{Function(optionalTagName, ShortCircuit|NonError, optionalValueValidator)}}, nil
+}
+
+var (
+	forbiddenValueValidator   = types.Name{Package: libValidationPkg, Name: "ForbiddenValue"}
+	forbiddenPointerValidator = types.Name{Package: libValidationPkg, Name: "ForbiddenPointer"}
+	forbiddenSliceValidator   = types.Name{Package: libValidationPkg, Name: "ForbiddenSlice"}
+	forbiddenMapValidator     = types.Name{Package: libValidationPkg, Name: "ForbiddenMap"}
+)
+
+// TODO: It might be valuable to have a string payload for when forbidden is
+// conditional (e.g. forbidden when <option> is disabled).
+func (requirednessTagValidator) doForbidden(context Context) (Validations, error) {
+	// Forbidden is weird.  Each of these emits two checks, which are polar
+	// opposites.  If the field fails the forbidden check, it will
+	// short-circuit and not run the optional check.  If it passes the
+	// forbidden check, it must not be specified, so it will "fail" the
+	// optional check and short-circuit (but without error).  Why?  For
+	// example, this prevents any further validation from trying to run on a
+	// nil pointer.
+	switch context.Type.Kind {
+	case types.Slice:
+		return Validations{
+			Functions: []FunctionGen{
+				Function(forbiddenTagName, ShortCircuit, forbiddenSliceValidator),
+				Function(forbiddenTagName, ShortCircuit|NonError, optionalSliceValidator),
+			},
+		}, nil
+	case types.Map:
+		return Validations{
+			Functions: []FunctionGen{
+				Function(forbiddenTagName, ShortCircuit, forbiddenMapValidator),
+				Function(forbiddenTagName, ShortCircuit|NonError, optionalMapValidator),
+			},
+		}, nil
+	case types.Pointer:
+		return Validations{
+			Functions: []FunctionGen{
+				Function(forbiddenTagName, ShortCircuit, forbiddenPointerValidator),
+				Function(forbiddenTagName, ShortCircuit|NonError, optionalPointerValidator),
+			},
+		}, nil
+	case types.Struct:
+		// The +forbidden tag on a non-pointer struct is not supported.
+		// If you encounter this error and believe you have a valid use case
+		// for forbiddening a non-pointer struct, please let us know! We need
+		// to understand your scenario to determine if we need to adjust
+		// this behavior or provide alternative validation mechanisms.
+		return Validations{}, fmt.Errorf("non-pointer structs cannot use the %q tag", forbiddenTagName)
+	}
+	return Validations{
+		Functions: []FunctionGen{
+			Function(forbiddenTagName, ShortCircuit, forbiddenValueValidator),
+			Function(forbiddenTagName, ShortCircuit|NonError, optionalValueValidator),
+		},
+	}, nil
+}
+
+func (rtv requirednessTagValidator) Docs() TagDoc {
+	doc := TagDoc{
+		Tag:    rtv.TagName(),
+		Scopes: rtv.ValidScopes().UnsortedList(),
+	}
+
+	switch rtv.mode {
+	case requirednessRequired:
+		doc.Description = "Indicates that a field is optional to clients."
```

Comment:
`requirednessRequired` why is this one optional to clients?  backwards comment?

Reply:
You are correct, the docs for requirednessRequired and requirednessOptional are currently flipped. I've updated the lines here so they properly describe the +k8s:required tag as indicating a required field, and +k8s:optional as indicating an optional field. Thanks for catching this!

Reply:
+1, I believe this initially was due to a cut-paste error. This is resolved now with https://github.com/kubernetes/kubernetes/pull/130349/commits/314819ef353ccfaa1311ce750039bba8458d7a3a